compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp

```diff
@@ -0,0 +1,608 @@
+///   1. In batch mode, a single fusion command is issued and a topological sort
+///      is automatically computed by the fusion. Since this applies a single
+///      fusion, no interleaved canonicalization/cse/enabling occur and the
+///      resulting fusion may not be as good.
+///   2. In iterative mode, the user is responsible for providing the fusion
+///      order but interleaved canonicalization/cse/enabling occur which may
+///      result in better fusion results.
+///      TODO: Transform dialect op to apply topological sort and avoid the
+///      manual intervention.
+/// If `resultingFusedOpsHandles` is a non-null pointer, the fused operation are
+/// appended in order.
+// TODO: apply forwarding pattern.
+template <typename TilingTransformOp, typename TileOrNumThreadSpec,
+          typename BatchOrIterativeFusion, typename TileOrNumThreadKind>
+static Value tileAndFuseAndDistributeImpl(
+    ImplicitLocOpBuilder &b, Value rootH, ValueRange opsHToFuse,
+    std::conditional_t<std::is_same_v<TileOrNumThreadKind, TileOrNumThreadList>,
+                       ArrayRef<int64_t>, ArrayRef<OpFoldResult>>
+        tileSizesOrNumThreads,
+    ArrayAttr threadDimMapping, SmallVector<Value> *resultingFusedOpsHandles) {
+  auto tileToForeachOp = b.create<TilingTransformOp>(
+      rootH, tileSizesOrNumThreads, TileOrNumThreadSpec(), threadDimMapping);
+  Value foreachThreadH = tileToForeachOp.getForeachThreadOp();
+  if (std::is_same<BatchOrIterativeFusion, BatchFusionSpec>::value) {
+    Value mergedOpsH =
+        b.create<MergeHandlesOp>(opsHToFuse, /*deduplicate=*/true);
+    Value fusedH = b.create<FuseIntoContainingOp>(mergedOpsH, foreachThreadH);
+    (void)fusedH;
+    assert(!resultingFusedOpsHandles && "Handle needs unpacking");
+  } else {
+    for (Value h : opsHToFuse) {
+      Value fusedH = b.create<FuseIntoContainingOp>(h, foreachThreadH);
+      if (resultingFusedOpsHandles) resultingFusedOpsHandles->push_back(fusedH);
+    }
+  }
+  return foreachThreadH;
+}
+
+/// Call tileAndFuseAndDistributeImpl with ArrayRef<int64_t> tilesSizes.
+template <typename TilingTransformOp = TileToForeachThreadOp,
+          typename BatchOrIterativeFusion = BatchFusionSpec>
+static Value buildTFDWithTileSizes(
+    ImplicitLocOpBuilder &b, Value rootH, ValueRange opsHToFuse,
+    ArrayRef<int64_t> tileSizes, ArrayAttr threadDimMapping = {},
+    SmallVector<Value> *resultingFusedOpsHandles = nullptr) {
+  return tileAndFuseAndDistributeImpl<
+      TilingTransformOp, transform::TileSizesSpec, BatchOrIterativeFusion,
+      TileOrNumThreadList>(b, rootH, opsHToFuse, tileSizes, threadDimMapping,
+                           resultingFusedOpsHandles);
+}
+
+/// Call tileAndFuseAndDistributeImpl with a handle to multiple tilesSizes.
+template <typename TilingTransformOp = TileToForeachThreadOp,
+          typename BatchOrIterativeFusion = BatchFusionSpec>
+static Value buildTFDWithTileSizes(
```

Comment:
Templatization and operator overloads... I am no C++ expert, so I am uncomfortable giving specific feedback, but this seems a bit hard to follow for vanilla C++ programmers like me.

Reply:
I think `Template Metaprogramming` is good for some cases (like allowing flexible and performant code for std::tuple, std::function, etc). But this is not the case in `buildTFDWithTileSizes`. I looked into the difference and found that one works on `ArrayRef<int64_t>` and the other work on `Value`. Can we avoid template metaprogramming for this case? Throwing out random ideas

1. Can we merge the two `buildTFDWithTileSizes` function? , can we make it `ArrayRef<OpFoldResult>` and force the callers to pass `ArrayRef<OpFoldResult>`?
2. Can we add an Option-like struct (like `SCFTilingOptions`, so we can set up the tile sizes through the Option, and pass it to `buildTFDWithTileSizes`?